### PR TITLE
Add E2E configuration changes tests

### PR DESCRIPTION
**Pull request by sabinaaledort**
Create two address pools and request a specific address pool for each Loadbalancer service. Validate functionality.

Create two peers and validate each FRR container is paired with the cluster nodes.

### CODE REVIEW

File: e2etest/bgp_test.go [hunk 1]
```diff
@@ -390,6 +390,167 @@ var _ = ginkgo.Describe("BGP", func() {
...
+
+	ginkgo.Context("validate configuration changes", func() {
+		ginkgo.AfterEach(func() {
+			// Clean previous configuration.
+			err := updateConfigMap(cs, configFile{})
+			framework.ExpectNoError(err)
+		})
+
+		ginkgo.It("request a specific address pool for each Loadbalancer service", func() {
+			ginkgo.By("configure first addresspool")
```
Comment (fedepaol) [inline e2etest/bgp_test.go:402]: Can we do the two iterations inside a loop?
That'd remove some code duplication

File: e2etest/bgp_test.go [hunk 2]
```diff
@@ -390,6 +390,167 @@ var _ = ginkgo.Describe("BGP", func() {
...
+
+			ginkgo.By("configure first service")
+			jig := e2eservice.NewTestJig(cs, namspace, "test-service1")
+			timeout := e2eservice.GetServiceLoadBalancerCreationTimeout(cs)
+			svc1, err := jig.CreateLoadBalancerService(timeout, func(svc *corev1.Service) {
+				tweakServicePort(svc)
+				svc.Spec.ExternalTrafficPolicy = corev1.ServiceExternalTrafficPolicyTypeCluster
+			})
+
+			framework.ExpectNoError(err)
```
Comment (fedepaol) [inline e2etest/bgp_test.go:448]: Nit: move this the line above

File: e2etest/layer2_test.go [hunk 3]
```diff
@@ -275,13 +275,154 @@ var _ = ginkgo.Describe("L2", func() {
...
+				firstAddressPool,
+			},
+		}
+
+		err := updateConfigMap(cs, configData)
+		framework.ExpectNoError(err)
+
+		namspace := f.Namespace.Name
+
+		ginkgo.By("configure first service")
```
Comment (fedepaol) [inline e2etest/layer2_test.go:303]: same as above. Can we have both inside a loop?

File: e2etest/bgp_test.go [hunk 4]
```diff
@@ -390,6 +390,130 @@ var _ = ginkgo.Describe("BGP", func() {
...
+					framework.ExpectEqual(ip, servicesIngressIP[j])
+
+					ginkgo.By(fmt.Sprintf("checking connectivity of service %d to its external VIP", j+1))
+					for _, c := range frrContainers {
+						validateService(cs, services[j], allNodes.Items, c)
+					}
+				}
+			}
+		})
+		ginkgo.It("configure peers one by one and validate FRR paired with nodes", func() {
```
Comment (fedepaol) [inline e2etest/bgp_test.go:490]: nit: can you add a space?

File: e2etest/layer2_test.go [hunk 5]
```diff
@@ -275,13 +275,102 @@ var _ = ginkgo.Describe("L2", func() {
...
+		}
+	},
+		table.Entry("IPV4", &ipv4ServiceRange),
+		table.Entry("IPV6", &ipv6ServiceRange))
 })
 
-func firstIPFromRange(ipRange string) string {
-	cidr, err := internalconfig.ParseCIDR(ipRange)
+func getIPFromRangeByIndex(ipRange string, index int) string {
+	cidrs, err := internalconfig.ParseCIDR(ipRange)
```
Comment (fedepaol) [inline e2etest/layer2_test.go:352]: why is not enough to pick the first cidr anymore?
IIUC. by doing this we go over all the possible cidrs related to the range, which is different from picking the first and the nth item of the same cidr.

File: e2etest/bgp_test.go [hunk 6]
```diff
@@ -390,6 +390,130 @@ var _ = ginkgo.Describe("BGP", func() {
...
 	})
+
+	ginkgo.Context("validate configuration changes", func() {
+		ginkgo.AfterEach(func() {
+			// Clean previous configuration.
+			err := updateConfigMap(cs, configFile{})
+			framework.ExpectNoError(err)
+		})
+
+		ginkgo.It("request a specific address pool for each Loadbalancer service", func() {
```
Comment (fedepaol) [inline e2etest/bgp_test.go:401]: Can you change the description to something like: it should work after subsequent configuration updates (or something alogn that line)?

File: e2etest/bgp_test.go [hunk 7]
```diff
@@ -390,6 +390,130 @@ var _ = ginkgo.Describe("BGP", func() {
...
+				}
+				err := updateConfigMap(cs, configData)
+				framework.ExpectNoError(err)
+
+				for _, c := range frrContainers {
+					pairExternalFRRWithNodes(cs, c)
+				}
+
+				ginkgo.By(fmt.Sprintf("configure service number %d", i+1))
+				jig := e2eservice.NewTestJig(cs, namspace, fmt.Sprintf("test-service%d", i+1))
```
Comment (fedepaol) [inline e2etest/bgp_test.go:450]: Is there a reason why you can't reuse `createServiceWithBackend` ?

File: e2etest/layer2_test.go (hunk 5 above)
Comment (sabinaaledort) [reply e2etest/layer2_test.go:352]: for example, when passing the following range `172.18.0.6-172.18.0.16` to the ParseCIDR function we get the following cidrs:`[172.18.0.6/31 172.18.0.8/29 172.18.0.16/32]`
if we want to get the 5th ip in the range we can't use the first cidr as it has only two ips.

File: e2etest/bgp_test.go (hunk 7 above)
Comment (sabinaaledort) [reply e2etest/bgp_test.go:450]: i didn't want to add a name parameter to this function. do you think we should?

File: e2etest/layer2_test.go (hunk 5 above)
Comment (fedepaol) [reply e2etest/layer2_test.go:352]: so I think a better option is to pick the largest CIDR and use that as a range reference. Otherwise we (at extreme) risk picking the same ip because you jump from one CIDR to another

Comment (sabinaaledort) [reply e2etest/layer2_test.go:352]: but the cidrs are not overlapping so i can't get the same ip, or they might me overlapping?

File: e2etest/bgp_test.go (hunk 6 above)
Comment (sabinaaledort) [reply e2etest/bgp_test.go:401]: it sounds more like the context description, i wanted to be more specific here describing what exactly the test does.

File: e2etest/bgp_test.go (hunk 7 above)
Comment (fedepaol) [reply e2etest/bgp_test.go:450]: That doesn't sound too bad. Another option is to add a slice of optional modifier functions so all the other calling sites won't be affected. Something like 
`createServiceWithBackend(cs clientset.Interface, namespace string, policy corev1.ServiceExternalTrafficPolicyType, modifiers ...func(*corev1.Service))

We can even refactor the traffic policy parameter to use that.
See https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis for more details.

Either ways are fine, right now it's a bit hard to follow.

File: e2etest/bgp_test.go (hunk 6 above)
Comment (fedepaol) [reply e2etest/bgp_test.go:401]: Yes but the point of this test is to try to change two different configurations I think. Otherwise, I think it's not totally clear.

File: e2etest/layer2_test.go (hunk 5 above)
Comment (fedepaol) [reply e2etest/layer2_test.go:352]: ugh, I completely mis-understood what ParseCIDR does.Ignore what I just said, thanks for explaining.

File: e2etest/layer2_test.go [hunk 8]
```diff
@@ -275,13 +275,102 @@ var _ = ginkgo.Describe("L2", func() {
...
+		c = ipaddr.NewCursor([]ipaddr.Prefix{*ipaddr.NewPrefix(cidr)})
+		for j := i; j < index; j++ {
+			if c.Next() != nil {
+				err := c.Set(c.Next())
+				framework.ExpectNoError(err)
+				i++
+			}
+		}
+		if i == index {
+			break
```
Comment (fedepaol) [inline e2etest/layer2_test.go:366]: We can just return here and return an error out of the loop

File: e2etest/bgp_test.go [hunk 9]
```diff
@@ -464,18 +473,156 @@ var _ = ginkgo.Describe("BGP", func() {
 			),
 		)
 	})
+
+	ginkgo.Context("validate configuration changes", func() {
+		ginkgo.AfterEach(func() {
+			// Clean previous configuration.
+			err := updateConfigMap(cs, configFile{})
+			framework.ExpectNoError(err)
+		})
```
Comment (oribon) [inline e2etest/bgp_test.go:482]: Why is this necessary? The one defined in the beginning of the Describe isn't enough?

Comment (sabinaaledort) [reply e2etest/bgp_test.go:482]: That's an after each to we do a cleanup before the next test.

Comment (oribon) [reply e2etest/bgp_test.go:482]: but doesn't the first aftereach already takes care of that? https://github.com/metallb/metallb/blob/9a08426c6d753d0e9323dac60152ba8f123cb507/e2etest/bgp_test.go#L113

Comment (fedepaol) [reply e2etest/bgp_test.go:482]: It does

Comment (sabinaaledort) [reply e2etest/bgp_test.go:482]: right, i will remove it

File: e2etest/bgp_test.go [hunk 10]
```diff
@@ -465,18 +474,150 @@ var _ = ginkgo.Describe("BGP", func() {
...
+				}
+			}
+		},
+			table.Entry("IPV4", "ipv4"),
+			table.Entry("IPV6", "ipv6"))
+	})
 })
 
-func createServiceWithBackend(cs clientset.Interface, namespace string, policy corev1.ServiceExternalTrafficPolicyType) (*corev1.Service, *e2eservice.TestJig) {
+func createServiceWithBackend(cs clientset.Interface, namespace string, jigName string, config func(*corev1.Service)) (*corev1.Service, *e2eservice.TestJig) {
```
Comment (fedepaol) [inline e2etest/bgp_test.go:612]: Sorry I wasn't clear, my suggestion about the callback(s) (as a variadic args) was to be able to override the service without adding an extra parameter.
I see here that for the name we need an extra param anyway, so there's no point in moving to the function.

Comment (sabinaaledort) [reply e2etest/bgp_test.go:612]: i think it's more clean to have the function this way so we can set the service parameters. what do you think?

Comment (fedepaol) [reply e2etest/bgp_test.go:612]: I think it will be clear if we'll need to add more parameters (which might be the case when https://github.com/metallb/metallb/pull/1065 lands. Adding this to change always the single parameter is a bit of an overcomplication imo.

In anycase, since we'll need to do that, I think the calling site would benefit from two global `config func(*corev1.Service)`, one as SetClusterTrafficPolicy and the other as SetLocalTrafficPolicy so you can use those as parameters for this function.

**Comment by fedepaol** [comment]
@sabinaaledort looks good but needs rebase.
The other tests support dual stack too now, but I think it's not really necessary here since here we are testing how different config changes are handled.

**Comment by fedepaol** [comment]
LGTM, thanks!

**Comment by fedepaol** [comment]
Ugh, CI was green but this broke badly. We must force rebasing against main.
